mars/oscar/backends/ray/communication.py

```diff
@@ -0,0 +1,278 @@
+        await self._in_queue.put(message)
+        return await self._out_queue.get()
+
+
+@register_server
+class RayServer(Server):
+    __slots__ = '_closed', '_address', '_channels', '_tasks'
+
+    scheme = 'ray'
+
+    def __init__(self, address, channel_handler: Callable[[Channel], Coroutine] = None):
+        super().__init__(DEFAULT_DUMMY_RAY_ADDRESS, channel_handler)
+        self._address = address
+        self._closed = asyncio.Event()
+        self._channels: Dict[ChannelID, RayServerChannel] = dict()
+        self._tasks: Dict[ChannelID, asyncio.Task] = dict()
+
+    @classproperty
+    @implements(Server.client_type)
+    def client_type(self) -> Type["Client"]:
+        return RayClient
+
+    @property
+    @implements(Server.channel_type)
+    def channel_type(self) -> ChannelType:
+        return ChannelType.ray
+
+    @staticmethod
+    @implements(Server.create)
+    async def create(config: Dict) -> "RayServer":
+        config = config.copy()
+        address = config.pop('address', DEFAULT_DUMMY_RAY_ADDRESS)
+        handle_channel = config.pop('handle_channel')
+        if urlparse(address).scheme != RayServer.scheme:  # pragma: no cover
+            raise ValueError(f'Address for RayServer '
+                             f'should be starts with "ray://", '
+                             f'got {address}')
+        if config:  # pragma: no cover
+            raise TypeError(f'Creating RayServer got unexpected '
+                            f'arguments: {",".join(config)}')
+        server = RayServer(address, handle_channel)
+        return server
+
+    @implements(Server.start)
+    async def start(self):
+        # nothing needs to do for ray server
+        pass
+
+    @implements(Server.join)
+    async def join(self, timeout=None):
+        wait_coro = self._closed.wait()
+        try:
+            await asyncio.wait_for(wait_coro, timeout=timeout)
+        except (futures.TimeoutError, asyncio.TimeoutError):
+            pass
+
+    @implements(Server.on_connected)
+    async def on_connected(self, *args, **kwargs):
+        channel_id = args[0]
+        _, peer_dest_address, peer_channel_index = channel_id
+        channel = RayServerChannel(peer_dest_address, peer_channel_index, channel_id)
+        self._channels[channel_id] = channel
+        if kwargs:  # pragma: no cover
+            raise TypeError(f'{type(self).__name__} got unexpected '
+                            f'arguments: {",".join(kwargs)}')
+        await self.channel_handler(channel)
+
+    @implements(Server.stop)
+    async def stop(self):
+        self._closed.set()
+
+    @property
+    @implements(Server.stopped)
+    def stopped(self) -> bool:
+        return self._closed.is_set()
+
+    async def __on_ray_recv__(self, channel_id: ChannelID, message):
+        channel = self._channels.get(channel_id)
+        if not channel:
+            self._tasks[channel_id] = asyncio.create_task(self.on_connected(channel_id))
+        while not channel:
```

Comment:
It's dangerous to do the `while loop` to get the channel. If the `self.on_connected` raises an exception before set the channel to `self._channels`, then the loop will not break, because the `self.on_connected` is running as an asyncio task.

Two ways to solve this problem:
1. Break the loop if the `self._tasks[channel_id]` is not running.
2. Returns the channel from `self.on_connected`, then runs the `self.channel_handler` as an asyncio task in `self.on_connected`.

Reply:
The `on_connected` will run forever, you can't return channel from it

Reply:
Create channel outside seems better, just like `DummyServer`